oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchConnection.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.index.elasticsearch;
+
+import org.apache.http.HttpHost;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.client.RestHighLevelClient;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+/**
+ * This class represents an Elasticsearch Connection with the related <code>RestHighLevelClient</code>.
+ * As per Elasticsearch documentation: the client is thread-safe, there should be one instance per application and it
+ * must be closed when it is not needed anymore.
+ * https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/_changing_the_client_8217_s_initialization_code.html
+ *
+ * The getClient() initializes the rest client on the first call.
+ * Once close() is invoked this instance cannot be used anymore.
+ */
+public class ElasticsearchConnection implements Closeable {
+
+    protected static final String SCHEME_PROP = "elasticsearch.scheme";
+    protected static final String DEFAULT_SCHEME = "http";
+    protected static final String HOST_PROP = "elasticsearch.host";
+    protected static final String DEFAULT_HOST = "127.0.0.1";
+    protected static final String PORT_PROP = "elasticsearch.port";
+    protected static final int DEFAULT_PORT = 9200;
+
+    protected static final Supplier<ElasticsearchConnection> defaultConnection = () ->
+            new ElasticsearchConnection(DEFAULT_SCHEME, DEFAULT_HOST, DEFAULT_PORT);
+
+    private String scheme;
+    private String host;
+    private int port;
+
+    private volatile RestHighLevelClient client;
```

Comment:
Can this be non volatile as each thread will be accessing it and this won't be changing frequently.

Reply:
It's better to keep it volatile because of the double-check synchronized in `getClient`.